### ISSUE TITLE
PayIns.CreateCardDirect throws null reference exception

### ISSUE THREAD

**Issue by reddyBell** [Mangopay/mangopay2-net-sdk#18]
For some reason : MangoPayApi.PayIns.CreateCardDirect(...) throws a null reference exception  while the transaction seems to go just fine : 

Here's a log (sandbox environment) : 

Card registration : 

```
POST https://api.sandbox.mangopay.com/v2/my-app/CardRegistrations/13851264 HTTP/1.1
Host: api.sandbox.mangopay.com
Connection: keep-alive
Content-Length: 186
Pragma: no-cache
Cache-Control: no-cache
Origin: http://localhost:56898
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36
Content-type: application/x-www-form-urlencoded
Accept: */*
Referer: http://localhost:56898/index.html
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.8

Id=13851264&RegistrationData=data%3DWFSUMMF63bMOVTJI5pIqW2Wga8qJ1mPOOiAyRe37l4Wrh7me1LotdTX4l5oALFVJIeibKnNMtUe65WTUsKc4w3i4P5sQMcetYWR3iJoHyZtiTjfMtFT8ICKQ1UEeY8TdkXvOWuMugGhwJp0VppBukA
 
```

Response :

```
HTTP/1.1 200 OK
Cache-Control: no-cache
Pragma: no-cache
Content-Type: application/json; charset=utf-8
Expires: -1
Server: Leetchi
Access-Control-Allow-Origin: http://localhost:56898
Date: Tue, 07 Jun 2016 01:03:32 GMT
Content-Length: 623

{"Id":"13851264","Tag":null,"CreationDate":1465261411,"UserId":"13464662","AccessKey":"1X0m87dmM2LiwFgxPLBJ","PreregistrationData":"S8HjKhXaPXeNlzbaHMqIv-vHl38GEqrOQZSxxIjc_ZqsvTNfScbIfCNjAl8CT0BeS4wCy-yiraxeE65tmxOe8A","RegistrationData":"data=WFSUMMF63bMOVTJI5pIqW2Wga8qJ1mPOOiAyRe37l4Wrh7me1LotdTX4l5oALFVJIeibKnNMtUe65WTUsKc4w3i4P5sQMcetYWR3iJoHyZtiTjfMtFT8ICKQ1UEeY8TdkXvOWuMugGhwJp0VppBukA","CardId":"13851265","CardType":"CB_VISA_MASTERCARD","CardRegistrationURL":"https://homologation-webpayment.payline.com/webpayment/getToken","ResultCode":"000000","ResultMessage":"Success","Currency":"EUR","Status":"VALIDATED"}
```

Call to CreateCardDirect :

```
POST https://api.sandbox.mangopay.com/v2.01/my-app/payins/card/direct HTTP/1.1
Authorization: Bearer 13611e5b2f454e76a18bb7aeebc329e6
Accept: application/json, application/xml, text/json, text/x-json, text/javascript, text/xml, application/json
User-Agent: RestSharp/104.4.0.0
Content-Type: application/json
Host: api.sandbox.mangopay.com
Content-Length: 341
Accept-Encoding: gzip, deflate

{"AuthorId":"13464662","CreditedUserId":"12845567","DebitedFunds":{"Currency":"EUR","Amount":25200},"Fees":{"Currency":"EUR","Amount":0},"CreditedWalletId":"12845568","SecureModeReturnURL":"http://mywebsite.com/secure-payment-success","CardId":"13851265","SecureMode":"FORCE","CardType":null,"Tag":null}
```

Response :

```
HTTP/1.1 200 OK
Cache-Control: no-cache
Pragma: no-cache
Content-Type: application/json; charset=utf-8
Expires: -1
Server: Leetchi
Date: Tue, 07 Jun 2016 01:03:34 GMT
Content-Length: 802

{"Id":"13851266","Tag":null,"CreationDate":1465261412,"AuthorId":"13464662","CreditedUserId":"12845567","DebitedFunds":{"Currency":"EUR","Amount":25200},"CreditedFunds":{"Currency":"EUR","Amount":25200},"Fees":{"Currency":"EUR","Amount":0},"Status":"CREATED","ResultCode":null,"ResultMessage":null,"ExecutionDate":null,"Type":"PAYIN","Nature":"REGULAR","CreditedWalletId":"12845568","DebitedWalletId":null,"PaymentType":"CARD","ExecutionType":"DIRECT","SecureMode":"FORCE","CardId":"13851265","SecureModeReturnURL":"http://mywebsite.com/secure-payment-success?transactionId=13851266","SecureModeRedirectURL":"https://api.sandbox.mangopay.com:443/Redirect/ACSWithValidation?token=1c65ae3edc704a5a946bef27dff64e57","SecureModeNeeded":true,"StatementDescriptor":null}
```

As you can see the status is set to "Created", however CreateCardDirect throws the following exception instead of returning :

> System.NullReferenceException: Object reference not set to an instance of an object.
>    at MangoPay.SDK.Core.RestTool.DoRequest[U,T](String idempotencyKey, String urlMethod, Pagination pagination, T entity)
>    at MangoPay.SDK.Core.RestTool.Request[U,T](String idempotencyKey, String urlMethod, String requestType, Dictionary`2 requestData, Pagination pagination, T entity)
>    at MangoPay.SDK.Core.APIs.ApiBase.CreateObject[U,T](String idempotencyKey, MethodKey methodKey, T entity, String entityId, String secondEntityId)
>    at MangoPay.SDK.Core.APIs.ApiBase.CreateObject[U,T](String idempotencyKey, MethodKey methodKey, T entity, String entityId)
>    at MangoPay.SDK.Core.APIs.ApiBase.CreateObject[U,T](String idempotencyKey, MethodKey methodKey, T entity)
>    at MangoPay.SDK.Core.APIs.ApiPayIns.CreateCardDirect(String idempotencyKey, PayInCardDirectPostDTO payIn)
>    at MangoPay.SDK.Core.APIs.ApiPayIns.CreateCardDirect(PayInCardDirectPostDTO payIn)

This happens at a very bad time, we are in a hurry to deliver a product....

**Comment by reddyBell** [comment]
I have the latest nugget package installed : 1.0.83

**Comment by hobailey** [comment]
Hi @reddyBell, I'm sorry to hear you've been having trouble, and when in a hurry it doesn't help of course. 
Can you confirm whether this is something that has just started happening, or have you never tried this method before?

**Comment by reddyBell** [comment]
Thanks for your message, I never tried this method before (new customer)

**Comment by reddyBell** [comment]
Just a total guest : since you use reflection to populate the response entities, this could be related to the fact the response DTO is lacking the newly introduced StatementDescriptor but I'm not sure about that.

**Comment by hobailey** [comment]
Hmm, perhaps. This has indeed been added to the SDK but Nuget hasn't yet been updated. I'm deploying it now and will let you know when it's ready to update on your side.

**Comment by hobailey** [comment]
Hey @reddyBell, Nuget is now updated with the latest version of the SDK, so please try again and let me know how that goes.

**Comment by reddyBell** [comment]
Thanks for your prompt assistance, unfortunately I have just updated the NuGet package but the error is still there.

**Comment by hobailey** [comment]
Hmm, how strange. Could you double check your code against the [unit test](https://github.com/Mangopay/mangopay2-net-sdk/blob/master/MangoPay.SDK.Tests/BaseTest.cs#L247) for example since this is passing without any problems :-/

**Comment by reddyBell** [comment]
I just checked and my code does work in your unit test. I'm wondering if we are watching the same thing, are you 100% sure that the version on NuGet is the same version as the one on GitHub ? (I just downloaded a zip of the latest source).

**Comment by reddyBell** [comment]
Otherwise I don't quite see where the problem is. I did a line by line comparison with my code to check that the same parameters are being passed to your methods, and also inspected the transactions. The transactions are the same, but the unit test client does not throw any exception.

**Comment by reddyBell** [comment]
Ok I think I find where the issue might be, in the NuGet version the card registration is done at :+1:
 
https://api.sandbox.mangopay.com/v2/
And the Payin at : 
https://api.sandbox.mangopay.com/v2.01/

However in the test project, both calls are made at :
https://api.sandbox.mangopay.com/v2.01/

**Comment by reddyBell** [comment]
One day wasted but I found out where the issue was. I directly referenced the GitHub project from my application to debug the source code. 

When returning the PayIn, the JSON from your server contains : "SecureModeNeeded":true, while the PayinDto [is defined with](https://github.com/Mangopay/mangopay2-net-sdk/blob/master/MangoPay.SDK/Entities/GET/PayInCardDirectDTO.cs) :

```
/// <summary>Secure mode needed.</summary>
        public String SecureModeNeeded { get; set; }
```

This caused the JSON.NET serializer to emit a parsing exception since "true" is not escaped. Then, in RestTool.DoRequest<U,T> no null check is done after sending the RestSharp request :

```
IRestResponse<U> restResponse = client.Execute<U>(restRequest);
responseObject = restResponse.Data;
```

Due to the deserialization issue responseObject is null, causing the null reference exception I was facing.

The mystery is : why is the test project working and not our application. I don't have the time to investigate, but I'd be concerned if I were you. This could be beat you in production very badly at some point. Right now I have forked the library but I look forward to getting back to the NuGet package as soon as the issue is fixed.

This exercise is probably worth a welcome discount ;)
